ndonnx/_core/_impl.py

```diff
@@ -1003,32 +1013,93 @@ def _via_dtype(
-    for floating point types.
-
-    Raises TypeError if the provided arrays are neither.
+    """Like ``_via_dtype`` but chooses a dtype from the available that doesn't result in
+    loss.
+
+    Raises
+    ------
+    TypeError
+        For non-numerical types and if the type can't be safely casted to
+        any available type.
     """
     promoted_values = promote(*arrays)
 
     dtype = promoted_values[0].dtype
 
+    # For logging
+    available_types = [t for t in [int_dtype, float_dtype, uint_dtype] if t is not None]
+
     via_dtype: dtypes.CoreType
-    if isinstance(dtype, (dtypes.Integral, dtypes.NullableIntegral)) or dtype in (
+    if isinstance(dtype, (dtypes.Unsigned, dtypes.NullableUnsigned)):
+        if (
+            uint_dtype is not None
+            and ndx.iinfo(dtype).bits <= ndx.iinfo(uint_dtype).bits
+        ):
+            via_dtype = uint_dtype
+        elif int_dtype is not None and ndx.iinfo(dtype).bits <= (
+            ndx.iinfo(int_dtype).bits
+            if use_unsafe_uint_cast
+            else ndx.iinfo(int_dtype).bits - 1
+        ):
+            via_dtype = int_dtype
+        else:
+            raise TypeError(
+                f"Can't upcast unsigned type `{dtype}`. Available implementations are for `{*available_types,}`"
+            )
+    elif isinstance(dtype, (dtypes.Integral, dtypes.NullableIntegral)) or dtype in (
         dtypes.nbool,
         dtypes.bool,
     ):
-        via_dtype = dtypes.int64
+        if int_dtype is not None and ndx.iinfo(dtype).bits <= ndx.iinfo(int_dtype).bits:
+            via_dtype = int_dtype
+        else:
+            raise TypeError(
+                f"Can't upcast signed type `{dtype}`. Available implementations are for `{*available_types,}`"
+            )
     elif isinstance(dtype, (dtypes.Floating, dtypes.NullableFloating)):
-        via_dtype = dtypes.float64
+        if (
+            float_dtype is not None
+            and ndx.finfo(dtype).bits <= ndx.finfo(float_dtype).bits
+        ):
+            via_dtype = float_dtype
+        else:
+            raise TypeError(
+                f"Can't upcast float type `{dtype}`. Available implementations are for `{*available_types,}`"
+            )
     else:
         raise TypeError(f"Expected numerical data type, found {dtype}")
 
     return _variadic_op(arrays, fn, via_dtype, cast_return)
 
 
+def _via_i64_f64(
+    fn: Callable[..., _CoreArray], arrays: list[Array], *, cast_return=True
+) -> ndx.Array:
+    """Like ``_via_dtype`` but uses ``i64`` for integer or boolean types and ``float64``
+    for floating point types.
+
+    Raises TypeError if the provided arrays are neither.
+    """
+    return _via_upcast(
+        fn,
+        arrays,
+        int_dtype=dtypes.int64,
+        float_dtype=dtypes.float64,
+        use_unsafe_uint_cast=True,  # TODO this can cause overflow, we should set it to false and fix all uses
```

Comment:
Let's check this by triggering the Array API tests too.

Reply:
We should automate the kernel support / standard diffing process - in the meantime, it [looks](https://github.com/Quantco/ndonnx/actions/runs/10076357986/job/27856564619#step:5:2484) like `Less` and `LessEqual` don't support unsigned types (but for some reason `Greater` and `GreaterEqual` do). There are only 13 uses so we can follow up on this.